### PR TITLE
`EvaluatorValue` cleanup

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR does the following changes:
~1. Most `c10::optional` is replaced with `std::optional`~ Edit: split out in https://github.com/NVIDIA/Fuser/pull/509
2. `EvaluatorValue` is now `using EvaluatorValue = DynamicType<double, int64_t, bool>;`
3. `DynamicType` is nullable, so `optional<DynamicType>` no longer makes sense. I replaced them with `DynamicType`.

Most changes are mechanical, except that:

To achieve 2, `DynamicType` actually needs to be improved:
- `as<T>` is now defined only if `T` belongs to `Ts...`
- `cast<T>` is replaced with `operator T`, and it is defined if and only if there exist a `T1` in `Ts...` such that `static_cast<T>(T1)` is define.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
Looks like torchscript is broken by this PR ?!!! Strange. I will split the `optional` change out hopefully this will help narrowing down.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/type_traits.h [hunk 1]
```diff
@@ -610,3 +610,30 @@ static_assert(!any_check(
...
+namespace can_static_cast_impl {
+
+template <typename, typename, typename = void>
+struct CanStaticCast : std::false_type {};
+
+template <typename From, typename To>
+struct CanStaticCast<
+    From,
+    To,
+    std::void_t<decltype(static_cast<To>(std::declval<From>()))>>
```
Comment (naoyam) [inline csrc/type_traits.h:626]: Ha, this is interesting.

**Review by naoyam — APPROVED** [review]
LGTM

**Pull request merged**